File: prototype/src/plugin-loader.ts
import type { IErrorSink, IWorkspace, WorkspaceServices } from './contracts.js';
import { BulletinBoard } from './runtime.js';
import { detail } from './diagnostics.js';

export const DEFAULT_PLUGINS = [new URL('./plugins/echo.js', import.meta.url), new URL('./plugins/hex.js', import.meta.url)];
export async function loadPlugins(urls: readonly URL[], services: WorkspaceServices,
  board: BulletinBoard, errors: IErrorSink): Promise<void> {
  for (const url of urls) {
    try {
      const plugin: unknown = await import(url.href);
      if (!plugin || typeof plugin !== 'object' || !('apiVersion' in plugin) || plugin.apiVersion !== 1
        || !('createWorkspace' in plugin) || typeof plugin.createWorkspace !== 'function') {
        throw new Error('plugin requires apiVersion=1 and createWorkspace');
      }
      const workspace: unknown = plugin.createWorkspace(services);
      if (!workspace || typeof workspace !== 'object' || !('name' in workspace) || typeof workspace.name !== 'string'
        || !('Process' in workspace) || typeof workspace.Process !== 'function') throw new Error('invalid Workspace instance');
      board.Register(workspace as IWorkspace);
    } catch (error) {
      errors.Report({ code: 'PLUGIN_LOAD_FAILED', component: 'loader', detail: detail(error) });
    }
  }
}

File: src/Dsn.Host/Plugins/PluginLoader.cs
using System.Reflection;
using System.Runtime.Loader;
using Dsn.Contracts;

namespace Dsn.Host;

public sealed class PluginLoader
{
    private readonly List<AssemblyLoadContext> contexts = [];
    public void Load(string path, IWorkspaceRegistration registry, WorkspaceServices services, bool required = true, IFilterRegistration? filters = null)
    {
        try
        {
            if (!File.Exists(path)) throw new FileNotFoundException("Plugin assembly not found", path);
            var context = new PluginContext(Path.GetFullPath(path)); contexts.Add(context);
            var assembly = context.LoadFromAssemblyPath(Path.GetFullPath(path));
            var types = assembly.GetTypes().Where(t => !t.IsAbstract && typeof(IWorkspacePlugin).IsAssignableFrom(t)).ToArray();
            var filterTypes = assembly.GetTypes().Where(t => !t.IsAbstract && typeof(IFilterPlugin).IsAssignableFrom(t)).ToArray();
            if (types.Length == 0 && filterTypes.Length == 0) throw new InvalidOperationException("No plugin factories found");
            foreach (var type in filterTypes)
            {
                var plugin = (IFilterPlugin)Activator.CreateInstance(type)!;
                if (plugin.ApiVersion != 1 || filters is null || !filters.RegisterFilter(plugin.Type, plugin.Create))
                    throw new InvalidOperationException("Unsupported or duplicate filter plugin");
            }
            foreach (var type in types)
            {
                var plugin = (IWorkspacePlugin)Activator.CreateInstance(type)!;
                if (plugin.ApiVersion != 1) throw new InvalidOperationException("Unsupported plugin API");
                if (!registry.Register(plugin.Create(services)) && required) throw new InvalidOperationException("Required plugin registration rejected");
            }
        }
        catch (Exception e) { services.Errors.Report(new("PLUGIN_LOAD_FAILED", "host", Detail: e.GetType().Name)); if (required) throw; }
    }
    private sealed class PluginContext(string path) : AssemblyLoadContext
    {
        private readonly AssemblyDependencyResolver resolver = new(path);
        protected override Assembly? Load(AssemblyName name)
        {
            if (name.Name == typeof(IWorkspace).Assembly.GetName().Name) return typeof(IWorkspace).Assembly;
            var dependency = resolver.ResolveAssemblyToPath(name);
            return dependency is null ? null : LoadFromAssemblyPath(dependency);
        }
        protected override nint LoadUnmanagedDll(string name)
        {
            var dependency = resolver.ResolveUnmanagedDllToPath(name);
            return dependency is null ? 0 : LoadUnmanagedDllFromPath(dependency);
        }
    }
}
